client/session.go

```diff
@@ -718,15 +718,18 @@ func (s *session) newHostQueue(host topology.Host, topoMap topology.Map) hostQue
 }
 
 func (s *session) Write(
-	namespace, id string,
+	namespace, id ts.ID,
 	t time.Time,
 	value float64,
 	unit xtime.Unit,
 	annotation []byte,
 ) error {
-	w := s.writeAttemptPool.Get()
-	w.args.namespace, w.args.id =
-		namespace, id
+	var (
+		w    = s.writeAttemptPool.Get()
+		nsID = s.idPool.Clone(namespace)
+		tsID = s.idPool.Clone(id)
```

Comment:
Hm, I think these will get finalized by the first write attempt at the end of the `writeState`.

A retry would probably cause a use after free.  Might need to clone it for every write attempt (simplest fix)?

Reply:
Yikes, nice catch. Added a test and fixed with by making a copy per writeState in the next revision.